File: visa-payment-go/internal/services/transaction_service.go
package services

import (
	"errors"
	"strings"
	"time"
	"visa-payment-service/internal/dto"
	"visa-payment-service/internal/models"
	"visa-payment-service/internal/repository"
)

type TransactionService struct {
	transactionRepo   *repository.TransactionRepository
	accountRepo       *repository.AccountRepository
	operationTypeRepo *repository.OperationTypeRepository
}

func NewTransactionService(
	transactionRepo *repository.TransactionRepository,
	accountRepo *repository.AccountRepository,
	operationTypeRepo *repository.OperationTypeRepository,
) *TransactionService {
	return &TransactionService{
		transactionRepo:   transactionRepo,
		accountRepo:       accountRepo,
		operationTypeRepo: operationTypeRepo,
	}
}

func (s *TransactionService) CreateTransaction(req *dto.TransactionRequest) (*models.Transaction, error) {
	// Validate account exists
	account, err := s.accountRepo.GetByID(req.AccountID)
	if err != nil {
		return nil, errors.New("account not found")
	}

	// Validate operation type exists
	operationType, err := s.operationTypeRepo.GetByID(req.OperationTypeID)
	if err != nil {
		return nil, errors.New("operation type not found")
	}

	// Apply business rules for amount sign based on operation type
	amount := req.Amount
	operationDescription := strings.ToLower(operationType.Description)

	// Purchase and withdrawal transactions should be negative
	if strings.Contains(operationDescription, "purchase") || strings.Contains(operationDescription, "withdrawal") {
		amount = -amount
	}
	// Credit voucher transactions should be positive (already positive from validation)

	transaction := &models.Transaction{
		AccountID:       req.AccountID,
		OperationTypeID: req.OperationTypeID,
		Amount:          amount,
		EventDate:       time.Now(),
		Account:         *account,
		OperationType:   *operationType,
	}

	err = s.transactionRepo.Create(transaction)
	if err != nil {
		return nil, err
	}

	// Reload with relations
	transaction, err = s.transactionRepo.GetByID(transaction.ID)
	if err != nil {
		return nil, err
	}

	return transaction, nil
}

File: visa-payment-java/src/main/java/com/visa/paymentservice/service/TransactionService.java
package com.visa.paymentservice.service;

import com.visa.paymentservice.dto.TransactionRequest;
import com.visa.paymentservice.model.Account;
import com.visa.paymentservice.model.OperationType;
import com.visa.paymentservice.model.Transaction;
import com.visa.paymentservice.repository.AccountRepository;
import com.visa.paymentservice.repository.OperationTypeRepository;
import com.visa.paymentservice.repository.TransactionRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.math.BigDecimal;
import java.util.Optional;

@Service
@Transactional
public class TransactionService {
    
    @Autowired
    private TransactionRepository transactionRepository;
    
    @Autowired
    private AccountRepository accountRepository;
    
    @Autowired
    private OperationTypeRepository operationTypeRepository;
    
    public Transaction createTransaction(TransactionRequest request) {
        // Validate account exists
        Optional<Account> accountOpt = accountRepository.findById(request.getAccountId());
        if (accountOpt.isEmpty()) {
            throw new IllegalArgumentException("Account with ID " + request.getAccountId() + " not found");
        }
        
        // Validate operation type exists
        Optional<OperationType> operationTypeOpt = operationTypeRepository.findById(request.getOperationTypeId());
        if (operationTypeOpt.isEmpty()) {
            throw new IllegalArgumentException("Operation type with ID " + request.getOperationTypeId() + " not found");
        }
        
        Account account = accountOpt.get();
        OperationType operationType = operationTypeOpt.get();
        
        // Apply business rules for amount sign based on operation type
        BigDecimal amount = request.getAmount();
        String operationDescription = operationType.getDescription().toLowerCase();
        
        // Purchase and withdrawal transactions should be negative
        if (operationDescription.contains("purchase") || operationDescription.contains("withdrawal")) {
            amount = amount.negate();
        }
        // Credit voucher transactions should be positive (already positive from validation)
        
        Transaction transaction = new Transaction(account, operationType, amount);
        return transactionRepository.save(transaction);
    }
}
